anvil/src/shell.rs

```diff
@@ -526,24 +629,101 @@ pub fn init_shell<BackendData: 'static>(display: &mut Display, log: ::slog::Logg
+                    surface,
+                    kind: ShellSurfaceKind::Fullscreen { output, .. },
+                } => {
+                    // NOTE: This is only one part of the solution. We can set the
+                    // location and configure size here, but the surface should be rendered fullscreen
+                    // independently from its buffer size
+                    let wl_surface = if let Some(surface) = surface.get_surface() {
+                        surface
+                    } else {
+                        // If there is no underlying surface just ignore the request
+                        return;
+                    };
+
+                    // Use the specified preferred output or else the current output the window
+                    // is shown or the primary output
+                    let output = output
+                        .map(|output| {
+                            shell_output_map
+                                .borrow()
+                                .find(&output, |_, geometry| (Some(output.clone()), geometry))
+                                .ok()
+                        })
+                        .flatten()
+                        .or_else(|| {
+                            let shell_window_map = shell_window_map.borrow();
+
+                            shell_window_map
+                                .find(wl_surface)
+                                .and_then(|kind| shell_window_map.location(&kind))
+                                .and_then(|position| {
+                                    shell_output_map
+                                        .borrow()
+                                        .find_by_position(position, |output, geometry| {
+                                            let mut window_output: Option<wl_output::WlOutput> = None;
+                                            output.with_output(wl_surface, |_, output| {
+                                                window_output = Some(output.to_owned());
+                                            });
+                                            (window_output, geometry)
+                                        })
+                                        .ok()
+                                })
+                        })
+                        .or_else(|| {
+                            shell_output_map
+                                .borrow()
+                                .with_primary(|output, geometry| {
+                                    let mut primary_output: Option<wl_output::WlOutput> = None;
+                                    output.with_output(wl_surface, |_, output| {
+                                        primary_output = Some(output.to_owned());
+                                    });
+                                    (primary_output, geometry)
+                                })
+                                .ok()
+                        });
```

Comment:
Can we de-duplicate at least this interesting construct and extract that into a helper function please? :smile:

Reply:
Yes, just forgot to move it into a function, good catch.

Reply:
I moved the code to a separate function within the shell and left a few comments what it does.